crates/m3/src/gadgets/mul.rs

```diff
@@ -398,12 +403,407 @@ impl MulUU64 {
+
+		let out_high = table.add_computed("out_high", pack_fp(out_high_bits));
+		let out_low = table.add_computed("out_low", pack_fp(out_low_bits));
+
+		Self {
+			mul_inner,
+			x_in_bits,
+			y_in_bits,
+			x_abs_value,
+			out_signed_value,
+			x_abs_bits,
+			out_bits,
+			x_in,
+			y_in,
+			out_low,
+			out_high,
+		}
+	}
+
+	pub fn populate_with_inputs<P>(
+		&self,
+		index: &mut TableWitnessSegment<P>,
+		x_vals: impl IntoIterator<Item = B32>,
+		y_vals: impl IntoIterator<Item = B32>,
+	) -> Result<()>
+	where
+		P: PackedField<Scalar = B128> + PackedExtension<B1> + PackedExtension<B32>,
+	{
+		let mut mul_inner_x = Vec::new();
+		let mut mul_inner_y = Vec::new();
+		{
+			let mut x_in_bits = array_util::try_map(self.x_in_bits, |bit| index.get_mut(bit))?;
+			let mut y_in_bits = array_util::try_map(self.y_in_bits, |bit| index.get_mut(bit))?;
+			let mut out_bits = array_util::try_map(self.out_bits, |bit| index.get_mut(bit))?;
+			let mut x_abs_bits = array_util::try_map(self.x_abs_bits, |bit| index.get_mut(bit))?;
+
+			let mut x_in = index.get_mut(self.x_in)?;
+			let mut y_in = index.get_mut(self.y_in)?;
+			let mut out_low = index.get_mut(self.out_low)?;
+			let mut out_high = index.get_mut(self.out_high)?;
+
+			for (i, (x, y)) in x_vals.into_iter().zip(y_vals.into_iter()).enumerate() {
+				let x_i32 = i32::from_le_bytes(x.val().to_le_bytes());
+				let x_abs = B32::new(abs(x_i32));
+				let res = x_i32 as i64 * y.val() as i64;
+				let res_high = B32::new((res >> 32) as u32);
+				let res_low = B32::new(res as u32);
+				mul_inner_x.push(x_abs);
+				mul_inner_y.push(y);
+				set_packed_slice(&mut x_in, i, x);
+				set_packed_slice(&mut y_in, i, y);
+				set_packed_slice(&mut out_low, i, res_low);
+				set_packed_slice(&mut out_high, i, res_high);
+
+				for bit_idx in 0..32 {
+					set_packed_slice(
+						&mut x_abs_bits[bit_idx],
+						i,
+						B1::from(u32::is_bit_set_at(x_abs, bit_idx)),
+					);
+					set_packed_slice(
+						&mut x_in_bits[bit_idx],
+						i,
+						B1::from(u32::is_bit_set_at(x, bit_idx)),
+					);
+					set_packed_slice(
+						&mut y_in_bits[bit_idx],
+						i,
+						B1::from(u32::is_bit_set_at(y, bit_idx)),
+					);
+					set_packed_slice(
+						&mut out_bits[bit_idx + 32],
+						i,
+						B1::from(u32::is_bit_set_at(res_high, bit_idx)),
+					);
+					set_packed_slice(
+						&mut out_bits[bit_idx],
+						i,
+						B1::from(u32::is_bit_set_at(res_low, bit_idx)),
+					);
+				}
+			}
+		}
+
+		self.x_abs_value.populate(index)?;
+		self.mul_inner.populate(index, mul_inner_x, mul_inner_y)?;
+		self.out_signed_value.populate(index)?;
+
+		Ok(())
+	}
+}
+
+/// Simple struct to convert to and from Two's complement representation based on bits. See [`SignConverter::new`]
+///
+/// NOTE: *We do not handle witness generation for the `converted_bits` and should be handled by caller*
+#[derive(Debug)]
+pub struct SignConverter<UPrimitive: UnsignedAddPrimitives, const BIT_LENGTH: usize> {
```

Comment:
is this just `Abs`? If so, wouldn't it be better, at least for discoverability sake, to name it this way?